Dockerfile

```diff
@@ -1,45 +1,33 @@
-# Use Rust 1.88 slim image for building
-FROM rust:1.88-slim-bookworm AS builder
+# -------------------
+# Stage 1: Build dependencies
+# -------------------
+FROM ghcr.io/astral-sh/uv:python3.14-bookworm-slim AS builder
 
-# Set working directory
-WORKDIR /usr/src/sparkth
+ENV UV_COMPILE_BYTECODE=1
+ENV UV_LINK_MODE=copy
+ENV UV_PYTHON_DOWNLOADS=0
 
-# Install build dependencies
-RUN apt-get update && apt-get install -y --no-install-recommends \
-    libssl-dev \
-    pkg-config \
-    && rm -rf /var/lib/apt/lists/*
+WORKDIR /app
 
-# Copy project files
-COPY . .
+RUN --mount=type=cache,target=/root/.cache/uv \
+    --mount=type=bind,source=uv.lock,target=uv.lock \
+    --mount=type=bind,source=pyproject.toml,target=pyproject.toml \
+    uv sync --frozen --no-install-project --no-dev
 
-# Define build profile (default: dist)
-ARG PROFILE=dist
+COPY . /app
```

Comment:
Let's avoid a rebuild of this step when there's a single change in the source code.

Reply:
This step is mandatory for `pyproject.toml` CLI entry points to work and allows us to import the app packages correctly. It also byte-compiles the source code for faster startup.

Modifying source code will technically re-run this final sync, but it will not trigger a re-install of external dependencies. Since the heavy dependencies are cached in the previous layer (via `--no-install-project`), this step only installs the local project, making the rebuild extremely fast.

https://github.com/astral-sh/uv-docker-example/blob/main/multistage.Dockerfile